lib/Sema/TypeCheckSwitchStmt.cpp

```diff
@@ -1023,6 +1022,39 @@ namespace {
         return eed->getAttrs().hasAttribute<DowngradeExhaustivityCheckAttr>();
       }
 
+      static void
+      decomposeTuple(TypeChecker &TC, const DeclContext *DC, TupleType *tuple,
+                     llvm::function_ref<void(ArrayRef<Space>)> processor) {
+        llvm::SmallVector<Space, 4> scratch;
+        decomposeTupleImpl(TC, DC, tuple->getElements(), 0, scratch, processor);
+      }
+
+      static void
+      decomposeTupleImpl(TypeChecker &TC, const DeclContext *DC,
+                         ArrayRef<TupleTypeElt> elements, unsigned startIndex,
+                         llvm::SmallVectorImpl<Space> &scratch,
+                         llvm::function_ref<void(ArrayRef<Space>)> processor) {
+        for (unsigned i = startIndex, n = elements.size(); i != n; ++i) {
+          llvm::SmallVector<Space, 4> spaces;
+
+          auto eltType = elements[i].getType();
+          if (canDecompose(eltType, DC)) {
+            decompose(TC, DC, eltType, spaces);
+          } else {
+            spaces.push_back(Space::forType(eltType, elements[i].getName()));
+          }
+
+          for (const auto &space : spaces) {
+            scratch.push_back(space);
+            decomposeTupleImpl(TC, DC, elements, i + 1, scratch, processor);
+            (void)scratch.pop_back_val();
+          }
+        }
+
+        if (startIndex == elements.size() && scratch.size() == elements.size())
```

Comment:
I agree with David; while I can tell that the purpose of the algorithm is to eagerly decompose tuple elements whenever a tuple is decomposed, I can't see how it does that, or whether it's correct for nested tuples, or what the significance of the sizes all matching up is. This needs more explanation in some form.